Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean

```diff
@@ -0,0 +1,174 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Comma.StructuredArrow
+
+/-!
+# Kan extensions
+
+The basic definitions for Kan extensions of functors is introduced in this file. Part of API
+is parallel to the definitions for bicategories (see `CategoryTheory.Bicategory.Kan.IsKan`).
+(The bicategory API cannot be used directly here because it would not allow the universe
+polymorphism which is necessary for some applications.)
+
+Given a natural transformation `α : L ⋙ F' ⟶ F`, we define the property
+`F'.IsRightKanExtension α` which expresses that `(F', α)` is a right Kan
+extension of `F` along `L`, i.e. that it is a terminal object in a
+category `RightExtension L F` of costructured arrows. The condition
+`F'.IsLeftKanExtension α` for `α : F ⟶ L ⋙ F'` is defined similarly.
+
+## TODO (@joelriou)
+
+* define a property `HasRightKanExtension L F` and related API
+* define a condition expressing that a functor is a pointwise Kan extension
+* refactor the file `CategoryTheory.Limits.KanExtension` using this new general API
+* define left/right derived functors as particular cases of Kan extensions
+
+## References
+https://ncatlab.org/nlab/show/Kan+extension
+
+-/
+
+namespace CategoryTheory
+
+open Category Limits
+
+namespace Functor
+
+variable {C D H : Type*} [Category C] [Category D] [Category H]
+
+/-- Given two functors `L : C ⥤ H` and `F : C ⥤ D`, this is the category of functors
+`F' : H ⥤ D` equipped with a natural transformation `L ⋙ F' ⟶ F`. -/
+abbrev RightExtension (L : C ⥤ H) (F : C ⥤ D) :=
+  CostructuredArrow ((whiskeringLeft C H D).obj L) F
```

Comment:
It may be a matter of preference, but we could write `whiskeringLeft ..` instead of `whiskeringLeft C H D` here.